src/metabase/email.clj

```diff
@@ -95,12 +98,66 @@
                   (assert (#{:tls :ssl :none :starttls} (keyword new-value))))
                 (setting/set-value-of-type! :keyword :email-smtp-security new-value)))
 
+(defsetting email-max-recipients-per-second
+  (deferred-tru "The maximum number of recipients, across emails, that can be sent per second.
+                In case an email has total number of recipients exceeding this limit, the throttling mechanism will be maxed-out but the exception will be silenced.")
+  :export?    true
+  :type       :integer
+  :visibility :settings-manager
+  :audit      :getter)
+
+(defn- make-email-throttler
+  [rate-limit]
+  (throttle/make-throttler
+   :email
+   :attempt-ttl-ms     1000
+   :initial-delay-ms   1000
+   :attempts-threshold rate-limit))
+
+(defonce ^:private email-throttler (when-let [rate-limit (email-max-recipients-per-second)]
+                                     (make-email-throttler rate-limit)))
+
+(defn check-email-throttle
+  "Check if the email throttler is enabled and if so, throttle the email sending based on the total number of recipients.
+
+  If an email has # of recipients greater than the rate limit:
+    a) Skip the throttle check if we haven't breached the rate limit yet.
+    b) If we have breached the rate limit, throttle like a normal email
+
+  We need a) to to avoid creating regression bugs due to the new throttling mechanism."
+  [email]
+  (when email-throttler
+    (when-some [recipients (into #{} (mapcat email) [:to :bcc])]
```

Comment:
Watch out for empty sets:

```suggestion
    (when-some [recipients (not-empty (into #{} (mapcat email) [:to :bcc]))]
```

TIL about `when-some` - its pretty cool! In this case the old school `when-let` is also fine :shrug

Reply:
hehe yeat that works too but I want to show off the `when-some` :D.

Reply:
you still need the `not-empty` though, empty set is still `some?`

let's add a test with no recipients?

Reply:
:) thought `(some? #{}) => false`. learned sth new today.